Add availability labels to combinations

Hlavtox opened the pull request:
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | See below
| Type?             | new feature
| Category?         | CO
| BC breaks?        | No BC breaks
| Deprecations?     | no
| Fixed ticket?     | Resolves #20536
| Related PRs       | Autougprade: https://github.com/PrestaShop/autoupgrade/pull/534
| How to test?      | See below
| Possible impacts? | -


#### Availability labels
- **NO BC BREAKS**
- Adds possibility to set specific availability labels to combinations. This availability label takes priority over the label set on product. So we can precisely specify, which combination are in suppliers warehouse and which are available only on request.
- Enable experimental product page and go edit any product combination. See that you can now set individual availability labels to any combination.
- You can also edit the labels via bulk edit tool.
- If no label is set on combination > product label is used. If no label on product > PS settings label is used. If no label in PS settings, nothing is shown.

#### ProductLazyArray refacto
- I refactored product lazy array a bit.
- Now there is one function that loads selected combination specific data. It retains original logic (a bit weird), but at least there is one source of truth in this class. _We can further improve this sometime, it contains different data in cart and product page._
- I simplified a bit the code that loads specific references, because it was loading other data from the combination. Now it will just take the 4 references and not care on anything else.
- I simplified code that loads reference to display.
- I improved the code that loads availability - we started it here originally (https://github.com/PrestaShop/PrestaShop/pull/25062/)

#### Database structure
- Lang table was added to Combination object - `product_attribute_lang`
- This table needs to be added in autoupgrade module and filled with empty entries, but according to my tests, it should work just fine even without the entries.

### How to test on existing install
- Create new table
```sql
CREATE TABLE `PREFIX_product_attribute_lang` (
  `id_product_attribute` int(10) unsigned NOT NULL,
  `id_lang` int(10) unsigned NOT NULL,
  `available_now` varchar(255) DEFAULT NULL,
  `available_later` varchar(255) DEFAULT NULL,
  PRIMARY KEY (`id_product_attribute`, `id_lang`)
) ENGINE='InnoDB' COLLATE 'utf8_general_ci';
```
And fill it for every language
```sql
INSERT INTO ps_product_attribute_lang (id_product_attribute, id_lang, available_now, available_later)
SELECT ps_product_attribute.id_product_attribute, 1, '', ''
FROM ps_product_attribute;
```

- Enable experimental product page. The new functionality will not be implemented for the legacy page.

#### Automated tests
- `tests/Integration/Adapter/Presenter/Product/ProductLazyArrayTest.php` was extended and improved to test all availability cases I could think of. This checks that it returns correct data - combination label, product label, ps config label.
- `tests/Integration/Behaviour/Features/Scenario/Product/Combination/update_combination_stock.feature` - added tests that check if combination labels can be edited and deleted properly.

Hlavtox committed:
Add combination availability labels

Hlavtox commented on src/Adapter/Presenter/Object/ObjectPresenter.php:
```diff
@@ -54,19 +53,8 @@ public function present($object)
             $presentedObject[$fieldName] = $object->{$fieldName};
         }
 
-        if ($object instanceof Product) {
```
This key was not used in the core anywhere, but I added it back to productcontroller just to be sure.

zuk3975 commented on classes/Combination.php:
```diff
@@ -94,6 +101,10 @@ class CombinationCore extends ObjectModel
             'available_date' => ['type' => self::TYPE_DATE, 'shop' => true, 'validate' => 'isDateFormat'],
+
+            /* Lang fields */
+            'available_now' => ['type' => self::TYPE_STRING, 'lang' => true, 'validate' => 'isGenericName', 'size' => 255],
+            'available_later' => ['type' => self::TYPE_STRING, 'lang' => true, 'validate' => 'IsGenericName', 'size' => 255],
```
It would be nice to define these "size" values in dedicated "CombinationSettings" class. (see example [ProductSettings](https://github.com/PrestaShop/PrestaShop/blob/develop/src/Core/Domain/Product/ProductSettings.php)). Then they can be reused in backOffice form types for constraints and if needed anywhere else (it would be the source of truth).

Basically these "{Foo}Settings" classes should be created for all such constants if there is no related ValueObject (if there is a value object, then the related constant stays there). Its kind of prestashop convention but idk if its written anywhere :smile: We have some examples with other domain classes - like SupplierSettings, AddressSettings (but probably we didn't went that far to also use those constants in related object model, altough we should have)

Hlavtox replied:
done

zuk3975 commented on controllers/front/ProductController.php:
```diff
@@ -1117,7 +1119,7 @@ public function getTemplateVarProduct()
         $product['quantity_wanted'] = $this->getRequiredQuantity($product);
         $product['extraContent'] = $extraContentFinder->addParams(['product' => $this->product])->present();
-
+        $product['ecotax_tax_inc'] = $this->product->getEcotax(null, true, true);
```
Is it related to PR? If not, maybe shouldn't add it even if it fixes something? Better to raise separate issue, because the PR has very clear scope and is already huge.

Hlavtox replied:
You are right, this is just "saw it, removed it". I can open another PR to clean this up.

Hlavtox replied:
Moved into separate PR https://github.com/PrestaShop/PrestaShop/pull/29462

zuk3975 commented on src/Adapter/Presenter/Object/ObjectPresenter.php:
```diff
@@ -54,19 +53,8 @@ public function present($object)
-
         $presentedObject['id'] = $object->id;
 
-        $mustRemove = ['deleted', 'active'];
-        foreach ($mustRemove as $fieldName) {
-            if (isset($presentedObject[$fieldName])) {
-                unset($presentedObject[$fieldName]);
-            }
-        }
-
```
same here, is it in scope of the PR?

Hlavtox replied:
I needed to use `active` property of the product in ProductLazyArray, this code just strips it... No need.

zuk3975 commented on src/Adapter/Presenter/Object/ObjectPresenter.php:
```diff
@@ -54,19 +53,8 @@ public function present($object)
             $presentedObject[$fieldName] = $object->{$fieldName};
         }
 
-        if ($object instanceof Product) {
-            $presentedObject['ecotax_tax_inc'] = $object->getEcotax(null, true, true);
-        }
-
```
Also doesn't seem to be in scope of PR,does it?

Hlavtox replied:
Same here - You are right, this is just "saw it, removed it". I can open another PR to clean this up.

Hlavtox replied:
Moved into separate PR https://github.com/PrestaShop/PrestaShop/pull/29462

zuk3975 commented on src/Adapter/Product/Combination/Validate/CombinationValidator.php:
```diff
@@ -90,6 +90,8 @@ private function validateStock(Combination $combination): void
         $this->validateCombinationProperty($combination, 'low_stock_alert', ProductConstraintException::INVALID_LOW_STOCK_ALERT);
         $this->validateCombinationProperty($combination, 'available_date', ProductConstraintException::INVALID_AVAILABLE_DATE);
+        $this->validateCombinationLocalizedProperty($combination, 'available_later', ProductConstraintException::INVALID_AVAILABLE_LATER);
+        $this->validateCombinationLocalizedProperty($combination, 'available_now', ProductConstraintException::INVALID_AVAILABLE_NOW);
```
All good here. Just I now see we forgot to add constraints to product available_now and available_later fields in [AvailabilityType](https://github.com/PrestaShop/PrestaShop/blob/develop/src/PrestaShopBundle/Form/Admin/Sell/Product/Stock/AvailabilityType.php#L88) and also "size" values should have been defined in ProductSettings and reused in Product.php. Forgot that too :smile: 
ping @jolelievre :smile:

zuk3975 replied:
https://github.com/PrestaShop/PrestaShop/issues/29454
created related issue for bugfix

zuk3975 commented on src/PrestaShopBundle/Form/Admin/Sell/Product/Combination/BulkCombinationStockType.php:
```diff
@@ -156,6 +157,23 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                 ],
                 'disabling_switch' => true,
             ])
+            ->add('available_now_label', TranslatableType::class, [
```
Add constraints (both in now and later label).  CombinationSettings class doesn't exist yet, refer to comment above for more info https://github.com/PrestaShop/PrestaShop/pull/29411/files#r956046238 . smth like:
```suggestion
            ->add('available_now_label', TranslatableType::class, [
                'options' => [
                    constraints' => [
                        new Length([
                            'max' => CombinationSettings::AVAILABLE_NOW_MAX_LENGTH
                        ]),
                        new TypedRegex(TypedRegex::TYPE_GENERIC_NAME),
                    ]
                ]
```

zuk3975 replied:
not sure if im not forgetting something and if bulkCombination form handles form constraint errors (but i think it should :thinking: )

Hlavtox replied:
done

zuk3975 commented on src/PrestaShopBundle/Form/Admin/Sell/Product/Combination/BulkCombinationStockType.php:
```diff
@@ -156,6 +157,23 @@ public function buildForm(FormBuilderInterface $builder, array $options)
+                'disabling_switch' => true,
+                'modify_all_shops' => true,
+            ])
+            ->add('available_later_label', TranslatableType::class, [
```
same for this

Hlavtox replied:
done

zuk3975 commented on src/PrestaShopBundle/Form/Admin/Sell/Product/Combination/CombinationStockType.php:
```diff
@@ -54,6 +56,21 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                     'placeholder' => 'YYYY-MM-DD',
                 ],
             ])
+            ->add('available_now_label', TranslatableType::class, [
```
and probably same here

Hlavtox replied:
done

zuk3975 commented on src/PrestaShopBundle/Form/Admin/Sell/Product/Combination/CombinationStockType.php:
```diff
@@ -54,6 +56,21 @@ public function buildForm(FormBuilderInterface $builder, array $options)
+                'required' => false,
+                'modify_all_shops' => true,
+            ])
+            ->add('available_later_label', TranslatableType::class, [
```
and here

Hlavtox replied:
done

zuk3975 commented on tests/Integration/Behaviour/Features/Context/Domain/Product/Combination/UpdateCombinationStockFeatureContext.php:
```diff
@@ -74,7 +74,9 @@ public function transformCombinationStock(TableNode $tableNode): CombinationStoc
             $dataRows['location'],
-            '' === $dataRows['available date'] ? null : new DateTime($dataRows['available date'])
+            '' === $dataRows['available date'] ? null : new DateTime($dataRows['available date']),
+            [],
+            []
```
thats not how you test a new feature :smile: 
Im afraid you will have to modify much more than that and add related scenrios to actually test it all works as expected.

Hlavtox replied:
@zuk3975 Yes, that was just to make tests work. 😄 I will maybe need some help on that 🙏

PS: Still better then deleting the test 🤣

Hlavtox replied:
done

kpodemski commented:
Thanks for your work, Daniel. There's a test for `ProductLazyArray` that checks the validity of the labels. Could you add valid tests to this?

It's here:
`tests/Integration/Adapter/Presenter/Product/ProductLazyArrayTest.php`

Hlavtox commented:
@zuk3975 

✅ Constraints added and working
![constraints](https://user-images.githubusercontent.com/6097524/188833905-89153273-108c-4927-9d42-86590938ed3c.jpg)

✅ Fixed Product::updateAttribute

atomiix commented on classes/Product.php:
```diff
@@ -2226,7 +2238,19 @@ public function updateAttribute(
                 'available_date' => null !== $available_date,
-            ]);
+            ];
+            if ($available_now !== null) {
```
```suggestion
            if (is_array($available_now)) {
```

atomiix commented on classes/Product.php:
```diff
@@ -2226,7 +2238,19 @@ public function updateAttribute(
+                    $fieldsToUpdate['available_now'][$id_lang] = true;
+                }
+            }
+            if ($available_later !== null) {
```
```suggestion
            if (is_array($available_later)) {
```

Hlavtox replied:
I am not sure about this, it can also be a string and it would be saved fine by the Object model 🤔 I think

atomiix commented on src/Adapter/Presenter/Product/ProductLazyArray.php:
```diff
@@ -541,49 +541,57 @@ public function getMainVariants()
-                'ean13' => $this->product['ean13'] ?? false,
-                'mpn' => $this->product['mpn'] ?? false,
-            ];
+        $data = array_slice($this->product['attributes'], 0);
+
+        return isset($data[0]) ? $data[0] : null;
```
As `$this->product['attributes']` cannot be empty because you checked it before, you can simply do:
```suggestion

        return reset($this->product['attributes'])
```

Hlavtox replied:
done

zuk3975 commented on src/Adapter/Presenter/Product/ProductLazyArray.php:
```diff
@@ -382,22 +379,25 @@ public function getGroupedFeatures()
-            $seoAvailability .= 'InStock';
-        } elseif ($this->product['quantity'] <= 0 && $this->product['allow_oosp']) {
-            $seoAvailability .= 'PreOrder';
+        // This case doesn't show by default, but if somebody modified Prestashop to keep
+        // discontinued products displayed, he will have proper structured data
+        if ($this->product['active'] === 0) {
+            return 'https://schema.org/Discontinued';
+        // Product in stock OR stock disabled (we have everything in stock)
+        } elseif ($this->product['quantity'] > 0 || !$this->configuration->get('PS_STOCK_MANAGEMENT')) {
+            return 'https://schema.org/InStock';
+        // All other cases (even if product is available for order) is OutOfStock
```
I don't think its related to PR scope.

Hlavtox replied:
Removed and moved to separate PR - https://github.com/PrestaShop/PrestaShop/pull/29570

zuk3975 commented:
all good for me except this https://github.com/PrestaShop/PrestaShop/pull/29411/files#r964826711
and behats must be adjusted correctly with at least couple default scenarios (now I see it only avoids errors providing empty labels in tests)

Hlavtox commented:
Hi @zuk3975 @atomiix @kpodemski 

- I added behaviour tests for saving and deleting the availability labels from combination.
- I added tests to `tests/Integration/Adapter/Presenter/Product/ProductLazyArrayTest.php` to test proper output of lazy array and the changes I made to it. Maybe we can (later) further extend on testing the specific references also. I have put some fake combination data in the label tests. If it gets reused, I will move it to class atribute and reuse it.
- One thing I want to ask about is `Product::updateAttribute()` function. 
  - This function is not used in the new product page anymore. It's used on the old page and maybe some other places and modules.
  - This function should theoretically accept both strings and arrays for the labels. We should reflect this in setFieldsToUpdate? If `available_later` is an array `id_lang =>value`, keep the current code. If it is a string, add `'available_later' => true`? 🤔

zuk3975 commented on classes/Product.php:
```diff
@@ -2212,12 +2222,14 @@ public function updateAttribute(
         $low_stock_alert = false,
-        $mpn = null
+        $mpn = null,
+        $available_now = [],
+        $available_later = []
```
these fields missing in docblock

zuk3975 commented on classes/Product.php:
```diff
@@ -2226,7 +2238,19 @@ public function updateAttribute(
                 'available_date' => null !== $available_date,
-            ]);
+            ];
+            if (is_array($available_now)) {
```
I think because these are new fields, you could strictly type them as array in the method and you wouldn't need these "is_array" checks.

zuk3975 commented on classes/Product.php:
```diff
@@ -2324,7 +2350,9 @@ public function addAttribute(
-        $mpn = null
+        $mpn = null,
+        $available_now = [],
+        $available_later = []
```
also I wonder if these should be `array` or `null|array`. If you use array and lets say you want to empty the values in all languages, you would have to specify each language as empty string. Whilst if using null|array we could identify if we need to change them or not. e.g. if its null we dont change anything, but if its empty array we delete them in all langs.
Im not sure about this, maybe @jolelievre can help with this?

Hlavtox replied:
Should be solved now exactly as you described. 👍

zuk3975 commented on classes/Product.php:
```diff
@@ -24,6 +24,7 @@
  * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
  */
 
+use Intervention\HttpAuth\Token\NullToken;
```
whats this?

zuk3975 commented on classes/Product.php:
```diff
@@ -2226,7 +2244,23 @@ public function updateAttribute(
+                    $fieldsToUpdate['available_now'][$id_lang] = true;
+                }
+            }
+            if (is_array($available_now)) {
```
```suggestion
            if (is_array($available_later)) {
```

Hlavtox replied:
fixed

zuk3975 reviewed (dismissed):
ok for me :heavy_check_mark:

l-delin reviewed (dismissed):
Checked wording, nothing to edit

Hlavtox commented:
@PrestaShop/prestashop-maintainers 
Hi guys, I updated the PR a bit:
- rebased
- applied some simplifications from our discussion in weekly meeting
- removed BC breaks from the PR to reduce it's scope
___________

@l-delin 
I need one more wording check, please 🙏

l-delin commented:
Wording check has been done ✅ removing label

marsaldev reviewed (dismissed):
The code seems fine to me

HanaRebaiQA reviewed (commented):
Hello @Hlavtox 

I have checked your PR. I found these issues : 

**Issue 1** : The global label which is setted in PS product settings label (eg. out of stock) is not taken into consideration when there is the in stock putted in the product label : It was Ok without PR

1. Edit any combination (eg. Out of stock behavior)
2. Do not set individual labels for both in-stock and out-of-stock behaviors
3. Set only the in-stock label in the product labels 
4. Put the out-of-stock label field empty in the product labels 
5. Set the in-stock and the out of stock in the PS product settings 
6. Go to FO > Check this combination 
7. See error > The out-of-stock label of the PS product settings is not displayed
8. Come back to products > edit the product with combinations > delete the in-stock label from the product labels 
9. Go to FO > refresh the page 
10. See that the label is well displayed under the combination with the out of stock label which is setted in the PS product settings 


https://user-images.githubusercontent.com/92912932/200291138-33fca4cb-4a90-419d-9e04-9d1ac82088fc.mp4


**Issue 2** : A bad display in the drop-down list for the languages with different translations on these pages : 

- When editing the combination 
- When using the bulk action to edit combinations 


![image](https://user-images.githubusercontent.com/92912932/200292390-02a50d40-4aad-4a93-9f02-43035f335e6f.png)

![image](https://user-images.githubusercontent.com/92912932/200292410-d8470801-09ea-4f05-a592-927c0eea961d.png)

![image](https://user-images.githubusercontent.com/92912932/200292430-3222bd3f-6bc1-4aec-8c16-3471de8c0cba.png)

https://user-images.githubusercontent.com/92912932/200292452-acd96aba-f3d1-42b0-a58d-5a36ecd85f56.mp4

Could you please check it ?

Thank You!

Hlavtox committed:
Fix typo

Hlavtox commented:
Hi @HanaRebaiQA, 

Issue 1 is fixed, tested, working. ✅

Issue 2 is not really related to this PR, but rather a global problem which should be solved in a different PR.

Hlavtox committed:
Align language switches to the right to prevent overflow

Hlavtox committed:
Fix dropdowns underflowing

Hlavtox commented:
@HanaRebaiQA Both issues fixed. ✅

For overflow to the right, I changed the language switch to be right aligned. This issue could occur also on other places maybe, it won't now.

For overflow to the bottom, it was because default boostrap styling had overflow: hidden on cards in accordions. So I changed this for accordions inside our form.

![win](https://user-images.githubusercontent.com/6097524/200817047-6229045d-5a58-4236-bb7e-47eaa4dda17b.jpg)

Hlavtox committed:
Fix stylelint

marsaldev commented:
@HanaRebaiQA need to be retested?

HanaRebaiQA commented:
Hello @marsaldev 

Yes of course after the changes done by @Hlavtox 
We are waiting the `Waiting for QA `label

Hlavtox commented:
@HanaRebaiQA I pinged you 5 days ago that it's ready to be retested, you don't have to wait for the label I think 😄

marsaldev commented:
> @HanaRebaiQA I pinged you 5 days ago that it's ready to be retested, you don't have to wait for the label I think 😄

Yeah, but before the QA test the PRs needs to be approved by 2 person 😊 , 5 days ago it wasn't 😊

HanaRebaiQA commented:
Hello @Hlavtox 

I have checked again your PR. i found this issue :

HanaRebaiQA reviewed (commented):
Hello @Hlavtox 

I have checked again your PR. I found these issues : 

**Issue 1:** Cannot add labels to combinations when editing with bulk action

1. Enable experimental fetaures 
3. Go to products V2 > edit a product with combination
4. Edit the combination with bulk action 
5. In `in-stock` and `out-stock `label > enable to toggle > Use the dropdown to change languages and add new labels for it
6. See error > Cannot choose another language!

https://user-images.githubusercontent.com/92912932/201903525-703788ea-d59d-4475-9184-f70c3a842288.mp4



**With dropdown in edit combination and edit labels in product and in products settings : OK**

https://user-images.githubusercontent.com/92912932/201903287-22f1d8a7-5fa1-4a5e-97b5-f6523262a46c.mp4



**Issue 2 :** When setting new labels in` in stock` and `out stock` labels with edit combinations, those labels are not displayed when editing the same combination with bulk action


https://user-images.githubusercontent.com/92912932/201903614-6aaf5643-49e8-4216-af01-246952967d3e.mp4


Could you please check it ?

Thank You!

Hlavtox committed:
Initialize language switcher in modal

Hlavtox commented:
Hi @HanaRebaiQA

Issue 1 is now fixed, the language switch was not initialized. After cloning the PR rebuild the assets to make it work.

https://user-images.githubusercontent.com/6097524/201916393-d10a887a-c62b-42ea-83ba-3748b1a7f414.mp4

Issue 2 is normal behavior. :-) When you bulk edit it, it doesn't show no values. It's for editing multiple ones at once, what value it should show if you selected 2 combinations? 😄 See that stock location behaves the same way.


https://user-images.githubusercontent.com/6097524/201916596-b7ac5d79-04df-48fd-8e93-eec5349f5f25.mp4

HanaRebaiQA commented:
ping @MatShir What do you think about issue2?

zuk3975 commented:
@HanaRebaiQA as @Hlavtox mentioned its expected behavior. It is a bulk action designed for multiple combinations update at once, its not expected to show any data because there are multiple different combinations with a different data.

MatShir commented:
Hi @HanaRebaiQA, 
Thanks for the tests, to confirm it's the normal behavior not to display any data of the products in the bulk. If you check the other value, it doesn't correspond to product data, but only default data like 0 quantity for ex.

If there are no more bugs, does it mean that we can merge the PR? @HanaRebaiQA :)

jolelievre commented on src/Adapter/Presenter/Product/ProductLazyArray.php:
```diff
@@ -541,49 +538,55 @@ public function getMainVariants()
 
-        if (empty($specificReferences)) {
-            $specificReferences = null;
+            $specificReferences[$this->getTranslatedKey($key)] = $value;
```
I believe this line will never be reached since you use `continue` in the last `else` statement

Hlavtox replied:
It will, I improved the code for readability. ;-)

jolelievre commented on src/PrestaShopBundle/Form/Admin/Sell/Product/Combination/BulkCombinationStockType.php:
```diff
@@ -156,6 +160,49 @@ public function buildForm(FormBuilderInterface $builder, array $options)
+                'label' => $this->trans('Label when in stock', 'Admin.Catalog.Feature'),
+                'required' => false,
+                'disabling_switch' => true,
+                'modify_all_shops' => true,
```
```suggestion
```
The new fields you added are not multishop (at least for now but maybe we should add this behaviour in a future PR) so this option should not be present, it will add the multishop checkbox in multishop mode even though it's not handled by the DB model.

Hlavtox replied:
Fixed

jolelievre commented on src/PrestaShopBundle/Form/Admin/Sell/Product/Combination/BulkCombinationStockType.php:
```diff
@@ -156,6 +160,49 @@ public function buildForm(FormBuilderInterface $builder, array $options)
+                ),
+                'required' => false,
+                'disabling_switch' => true,
+                'modify_all_shops' => true,
```
```suggestion
```

Hlavtox replied:
Fixed

jolelievre commented on src/PrestaShopBundle/Form/Admin/Sell/Product/Combination/CombinationStockType.php:
```diff
@@ -54,6 +59,47 @@ public function buildForm(FormBuilderInterface $builder, array $options)
+                'type' => TextType::class,
+                'label' => $this->trans('Label when in stock', 'Admin.Catalog.Feature'),
+                'required' => false,
+                'modify_all_shops' => true,
```
```suggestion
```

Hlavtox replied:
Fixed

jolelievre commented on src/PrestaShopBundle/Form/Admin/Sell/Product/Combination/CombinationStockType.php:
```diff
@@ -54,6 +59,47 @@ public function buildForm(FormBuilderInterface $builder, array $options)
+                    'Admin.Catalog.Feature'
+                ),
+                'required' => false,
+                'modify_all_shops' => true,
```
```suggestion
```

Hlavtox replied:
Fixed

jolelievre commented on src/PrestaShopBundle/Resources/views/Admin/TwigTemplateForm/bootstrap_4_layout.html.twig:
```diff
@@ -281,7 +281,7 @@
             </button>
 
-            <div class="dropdown-menu locale-dropdown-menu" aria-labelledby="{{ form.vars.id }}">
+            <div class="dropdown-menu dropdown-menu-right locale-dropdown-menu" aria-labelledby="{{ form.vars.id }}">
```
Be careful you are modifying the translatable inputs for the whole BO here, even though it fixes the issue you had in the bulk iframe (if I remember the issue correctly) Do we really need to apply this everywhere? Maybe this should be configurable via the `TranslatableType` option, to set this on a case by case?

jolelievre commented on src/PrestaShopBundle/Resources/views/Admin/TwigTemplateForm/prestashop_ui_kit_base.html.twig:
```diff
@@ -465,7 +465,7 @@
         </button>
 
-        <div class="dropdown-menu locale-dropdown-menu" aria-labelledby="{{ form.vars.id }}">
+        <div class="dropdown-menu dropdown-menu-right locale-dropdown-menu" aria-labelledby="{{ form.vars.id }}">
```
SAme comment as in the other form theme.

jolelievre reviewed (commented):
Thank you @Hlavtox 

Amazing work! Nothing much to say except a few details, and I had a question do you think these fields should be multishop as they are for the Product entity? Your new `product_attribute_lang` table is multilang but not multishop, although it can be added in another PR!

kpodemski commented:
@jolelievre I think you wanted to "Request changes"? This change in template files seems quite important

HanaRebaiQA commented:
Hello, 

Thank you @MatShir for your answer :) 
I should re-test it again after @Hlavtox fix.

jolelievre commented:
@kpodemski actually I'm not sure if it's worth blocking it, that's why I asked for more details

Hlavtox committed:
Remove multistore related keys from fields, they are not needed

Hlavtox committed:
Improve comments and variable naming for specific references

jolelievre reviewed (dismissed):
Thanks @Hlavtox

Hlavtox committed:
Remove extra isset

Hlavtox commented:
Guys, everything is done. So please last approval, sorry. 👍

@jolelievre @kpodemski @zuk3975

HanaRebaiQA reviewed (approved):
Hello @Hlavtox 

Thanks for all the fixed issues. I have checked again this PR and here is my feedback : 

- Check the automatic tests: OK
- Check all mentioned issues: OK
- Check when setting individuals labels in edit combinations: OK
- Check when setting labels with bulk action: OK
- Check with setting labels in the global product: OK
- Check setting labels in product settings: OK
- Check when setting labels only in `in-stock` label on the individual labels, global products, and product settings: OK
- Check when setting labels only in `out-of-stock `labels on the individual labels, global products, and product settings: OK
- Check with multistore: OK
- Check with different languages (English, French, Arabic, China): OK
- Check with different browsers (Chrome & Firefox): OK

https://user-images.githubusercontent.com/92912932/203079473-c1c6bb77-da14-435a-89eb-2c4cc35dd674.mp4


![image](https://user-images.githubusercontent.com/92912932/203077230-4c76634c-6acf-41d4-926c-b5b15d4cc43d.png)


So, it's QA :heavy_check_mark: 

Thank You!

kpodemski commented:
Thank you, @Hlavtox, and to all involved! Great stuff!

Hlavtox commented:
Thanks everyone! 🚀
